lib/cn/cn_tree.c

```diff
@@ -1965,20 +1961,262 @@ static merr_t
+    zspillingp = &znode->tn_ss_spilling;
+    atomic_inc_acq(zspillingp);
+    mutex_unlock(&tree->ct_ss_lock);
+    rmlock_runlock(lock);
+
+    ss_arr = calloc(CN_FANOUT_MAX, sizeof(*ss_arr));
+    if (!ss_arr)
+        return merr(ENOMEM);
+
+    i = 0;
+    while (1) {
+        struct route_node *rtnext;
+        struct subspill *ss;
+
+        rmlock_rlock(&tree->ct_lock, &lock);
+
+        rtnext = rtn ? route_node_next(rtn) : route_map_first_node(tree->ct_route_map);
+        if (!rtnext)
+            break;
+
+        n = route_node_tnode(rtnext);
+        if (!n->tn_route_node)
+            abort();
+
+        mutex_lock(&tree->ct_ss_lock);
+        if (n->tn_ss_splitting || n->tn_ss_joining < 0) {
+            ss = list_last_entry_or_null(&n->tn_ss_list, typeof(*ss), ss_link);
+
+
+            /* It was established upfront that the znode is not undergoing a spill or a join.
+             */
+            if (n->tn_nodeid != znode->tn_nodeid && (!ss || w->cw_sgen > ss->ss_sgen)) {
+                char *wmesg = n->tn_ss_splitting ? "spltwait" : "joinwait";
+
+                rmlock_runlock(lock);
+
+                atomic_inc(&tree->ct_rspill_slp);
+                cv_wait(&tree->ct_ss_cv, &tree->ct_ss_lock, wmesg);
+                atomic_dec(&tree->ct_rspill_slp);
+
+                mutex_unlock(&tree->ct_ss_lock);
+                continue;
+            }
+
+            ss = NULL;
+        }
+
+        /* Drop ref from previous subspill.
+         */
+        if (spillingp)
+            atomic_dec_rel(spillingp);
+
+        spillingp = &n->tn_ss_spilling;
+        atomic_inc_acq(spillingp);
+        mutex_unlock(&tree->ct_ss_lock);
+
+        rtn = rtnext;
+        rmlock_runlock(lock);
+
+        ss_arr[i].ss_work = w;
+        ss_arr[i].ss_sgen = w->cw_sgen;
+        ss_arr[i].ss_node = n;
+        ss_arr[i].ss_is_zspill = true;
+
+        if (n->tn_nodeid == znode->tn_nodeid) {
+            ss_arr[i].ss_added = true;
+            ss_arr[i].ss_zspill.zsp_src_list = first;
+        }
+
+        if (ss_arr[i].ss_sgen == atomic_read(&n->tn_sgen) + 1) {
+            err = cn_subspill_apply(&ss_arr[i]);
+            if (err)
+                break;
+
+            atomic_inc_rel(&n->tn_sgen);
+        } else {
+            atomic_inc(spillingp);
+            cn_subspill_enqueue(&ss_arr[i], n);
+        }
+
+        /* Apply subspills that are ready. */
+        while ((ss = cn_subspill_pop(n))) {
+            err = cn_subspill_apply(ss);
+            if (err) {
+                atomic_dec_rel(spillingp);
+                goto out;
+            }
+
+            atomic_inc_rel(&n->tn_sgen);
+            atomic_dec_rel(spillingp);
+        }
+
+        i++;
+    }
+
+    if (!err)
+        *zspilled = true;
+
+out:
+    rmlock_runlock(lock);
+    free(ss_arr);
+
+    if (spillingp)
+        atomic_dec_rel(spillingp);
+    if (zspillingp)
+        atomic_dec_rel(zspillingp);
+
+    if (err) {
+        if (merr_errno(err) != ESHUTDOWN)
```

Comment:
get rid of the first if statement

Reply:
You mean get rid of 'if (err) {' ???  Then it would call cn_node_spill_wait() if err was ESHUTDOWN, which puts a hidden requirement on cn_node_spill_wait() to know that the caller is in an ESHUTDOWN state and to always return ESHUTDOWN.
As written, cn_node_spill_wait() will return any current health error first, and if none, then will return ESHUTDOWN if cw_cancel_request is true.   So we should leave the err check...

Reply:
Oops I get it. I would just write `if (err && merr_errno(err) != ESHUTDOWN)`